credoai/modules/dataset_modules/dataset_base.py

```diff
@@ -61,4 +63,44 @@ def _make_pipe(self):
                [('scaler', StandardScaler()),
                 ('model', model)])
         return pipe
+
+    def _assess_proxy_in_feature_space(
+        self, excluded_columns=None, method="mutual_information"
+    ):
+        """Assesses features based on how much they serve as a proxy for a sensitive attribute
+        This method relies on features information only (label information is not used)
+
+        Parameters
+        ----------
+        excluded_columns : list[str]
+            List of columns to exclude from the assessment
+        method : str, optional
+            Method to use for quantifying the magnitude of proxy, by default 'mutual_information'
+            Choices:
+                'mutual_information' (the "amount of information" in normalized nats obtained about a feature by observing another feature)
+
+        Returns
+        -------
+        dict
+            Keys are feature names and values are their proxy scores
+            Score between 0.0 and 1.0
+        """
+        proxy_scores = {}
+        
+        if method == "mutual_information":
+            # Exclude features of type float
+            float_cols = list(self.X.select_dtypes(include=["float16", "float32", "float64"]))
```

Comment:
Why are we excluding `float`? You can estimate the information content between continuous and categorical variables. See [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.mutual_info_classif.html)

Reply:
it does not work like that. try some combinations of this:
```
df = pd.DataFrame({'race': ['white', 'white', 'white', 'white', 'white', 'white', 'black', 'asian', 'asian', 'asian', 'asian', 'asian'],
                    'gender': ['f', 'f', 'f', 'm', 'f', 'f', 'f', 'f', 'm', 'm', 'm', 'm'],
                  'experience': [0, 0.1, 0.2, 0.4, 0.5, 0.6, 0, 0.1, 0.2, 0.4, 0.5, 0.6],
                  'mark': [0, 0.3, 0.2, 0.4, 0.5, 0.6, 0.3, 0.1, 0.2, 0.4, 0.9, 0.6],
                  'zip code': [22203, 22205, 22203, 22203, 22203, 22203, 22204, 22204, 22204, 22204, 22203, 22204]})
```

Reply:
I believe the below works. You need to get the categorical variables into a coded form, and inform `mutual_info_classif`

```
df = pd.DataFrame({'race': ['white', 'white', 'white', 'white', 'white', 'white', 'black', 'asian', 'asian', 'asian', 'asian', 'asian'],
                    'gender': ['f', 'f', 'f', 'm', 'f', 'f', 'f', 'f', 'm', 'm', 'm', 'm'],
                  'experience': [0, 0.1, 0.2, 0.4, 0.5, 0.6, 0, 0.1, 0.2, 0.4, 0.5, 0.6],
                  'mark': [0, 0.3, 0.2, 0.4, 0.5, 0.6, 0.3, 0.1, 0.2, 0.4, 0.9, 0.6],
                  'zip code': [22203, 22205, 22203, 22203, 22203, 22203, 22204, 22204, 22204, 22204, 22203, 22204]})
# make longer, for sake of demo
df = pd.concat([df]*3, axis=0)

# convert cat columns to numeric to use
cat_columns = df.select_dtypes(include=['object']).columns
for col in cat_columns:
    df[col] = df[col].astype('category')
    df[f"{col}_codes"] = df[col].cat.codes

# drop categorical variables, just keep codes
processed_df = df.drop(cat_columns, axis=1)

# label which ones are categorical
cat_columns = ['zip code', 'race_codes', 'gender_codes']
discrete_features = [True if col in cat_columns else False 
                     for col in processed_df.columns]

mi= mutual_info_classif(processed_df, df['gender_codes'], 
                        discrete_features=discrete_features,random_state=42)
pd.Series(mi, index=processed_df.columns)
```

Reply:
This returns wrong/misleading results. MI of a variable with itself must be 1.0:

experience      0.153942
mark            0.241829
zip code        0.080515
race_codes      0.245412
**gender_codes    0.679193**

Reply:
With normalization
```
import pandas as pd
from sklearn.metrics import *
from sklearn.feature_selection import *

df = pd.DataFrame({'race': ['white', 'white', 'white', 'white', 'white', 'white', 'black', 'asian', 'asian', 'asian', 'asian', 'asian'],
                    'gender': ['f', 'f', 'f', 'm', 'f', 'f', 'f', 'f', 'm', 'm', 'm', 'm'],
                  'experience': [0, 0.1, 0.2, 0.4, 0.5, 0.6, 0, 0.1, 0.2, 0.4, 0.5, 0.6],
                  'mark': [0, 0.3, 0.2, 0.4, 0.5, 0.6, 0.3, 0.1, 0.2, 0.4, 0.9, 0.6],
                  'zip code': [22203, 22205, 22203, 22203, 22203, 22203, 22204, 22204, 22204, 22204, 22203, 22204]})
# make longer, for sake of demo
df = pd.concat([df]*3, axis=0)

# convert cat columns to numeric to use
cat_columns = df.select_dtypes(include=['object']).columns
for col in cat_columns:
    df[col] = df[col].astype('category')
    df[f"{col}_codes"] = df[col].cat.codes

# drop categorical variables, just keep codes
processed_df = df.drop(cat_columns, axis=1)

# label which ones are categorical
cat_columns = ['zip code', 'race_codes', 'gender_codes']
discrete_features = [True if col in cat_columns else False 
                     for col in processed_df.columns]

mi = mutual_info_classif(processed_df, df['gender_codes'], 
                        discrete_features=discrete_features,random_state=42)
out = pd.Series(mi, index=processed_df.columns)
normalized = out/out.max()
normalized
```